Call storage_service notifications only after keyspace schema changes are applied on all shards

bhalevy opened the pull request:
This series aims at hardening schema merges and preventing inconsistencies across shards by
updating the database shards before calling the notification callback.

As seen in #13137, we don't want to call the notifications on all shards in parallel while the database shards are in flux.

In addition, any error to update the keyspace will cause abort so not to leave the database shards in an inconsistent state .

Other changes optimize this path by:
- updating shard 0 first, to seed the effective_replication_map.
- executing `storage_service::keyspace_changed` only once, on shard 0 to prevent quadratic update of the token_metadata and e_r_m on every keyspace change.

Fixes #13137

tgrabiec commented on db/schema_tables.cc:
```diff
@@ -947,11 +948,23 @@ static future<> with_merge_lock(noncopyable_function<future<> ()> func) {
+    } catch (...) {
+        on_fatal_internal_error(slogger, fmt::format("Updating database schema on all shards failed: {}", std::current_exception()));
+    }
+}
```
This patch changes the behavior in a significant way. We will abort the node on failure instead of failing the schema merge. The title should communicate that.

tgrabiec replied:
The asymmetry between the way we handle shard 0 failure and non-shard-0 failure (throw vs abort) is not helpful in the more general scope. Even though this update throws with no side effects, prior calls to update_database() which completed already had side effects.

tgrabiec replied:
The decision how to handle that could be a separate PR. If erm is computed only on shard 0, we don't need this patch to ensure consistency.

bhalevy replied:
It's helpful for cross-shard consistency.
I.e. if the schema change was partially applied, at least all shards should be consistent with each other.
But indeed not consistent with the latest schema.
If it's an invalid state to be in, maybe we need to consider aborting after such an error, and let the schema be applied after restart,

bhalevy replied:
I agree.  I think it would be better to abort if we can't ensure consistency, rather than deal with broken state.  This way the database objects will be repopulated after restart.

bhalevy replied:
See `database: update_keyspace_on_all_shards: abort on error` in latest version

michoecho reviewed (changes requested):
Looks OK as far as I understand this area of code, but I'm upset by the lack of documentation. This is a concurrent algorithm (with distributed consequences) and it deserves more precision.
Mathematical rigor is the only valid way of working with concurrent algorithms — if a correctness patch for a concurrent algorithm uses the words "harden" or "prevent" in the description, something went wrong.

Aside from that, here's something I don't understand: why is the effective replication map recomputed both in `keyspace_changed --> update_pending_ranges --> replicate_to_all_cores` and in `keyspace::update_from`? Are there some circumstances where one is called, but the other isn't? Responsibilities are unclear here.

And for that matter, why is `update_pending_ranges` — a function nominally responsible for dealing with "pending ranges" (which I understand are token ranges with temporary extra write replicas due to a topology change) — responsible for updating effective replication maps after the replication strategy is changed?

michoecho commented on service/storage_service.cc:
```diff
@@ -4814,13 +4814,19 @@ future<> storage_service::update_topology_change_info(sstring reason, acquire_me
+        // FIXME: this is a serious error
+        // Currently migration_notifier::update_keyspace will ignore any exception we throw here anyhow
+        // but we need a way to propagate the error back, or abort instead.
+        slogger.error("Failed to update pending ranges for {}: {}. Ignored", reason, std::current_exception());
```
The patch doesn't "spread" it because the exception is also ignored (via `handle_exception`) before the patch.

bhalevy replied:
It was ignore before this change so I just preserved it.
I agree with you it's not OK to ignore it and it should be handled and propagated back up to schema merge.

bhalevy commented on db/schema_tables.cc:
```diff
@@ -947,11 +950,34 @@ static future<> with_merge_lock(noncopyable_function<future<> ()> func) {
+requires std::same_as<futurize_t<std::invoke_result_t<Func, replica::database&>>, future<>>
+static future<> update_database(sharded<replica::database>& db, Func func) {
+    using futurator = futurize<std::invoke_result_t<Func, replica::database&>>;
+    co_await futurator::invoke(func, db.local());
+    try {
+        co_await db.invoke_on_others(func);
```
It could be any shard for that matter, for the purpose of "seeding" a new effective replication map.
Other shards should find the new effective replication map, if one was created, and use it.

michoecho replied:
Ah, so this is just an optimization?

Please document that.

tgrabiec replied:
But those listeners are not only used for erm replication. The promise made by the notifier is that the listener is called on each shard. Some notification handlers can rely on the state on the local shard and not take it from shard 0. Some of them invalidate shard-local caches, for example.

If the state is taken from shard 0, it's better not to have each shard fetch the state from shard 0 (seed), but handle the notification only on shard 0 and have the handler replicate to other shards explicitly (like in replicate_to_all_cores()).

bhalevy replied:
the listeners are still notified on all shards, just in a 2nd pass.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1675/

bhalevy commented:
In v2 ([23323c7](https://github.com/scylladb/scylladb/commit/23323c7c0392d57672c78d32ed1435ad639f91e5)):
* focuses on keyspace operations (create, update, drop)
* added respective *_keyspace_on_all_shards static method to database
  - to take over responsibility for running the lower level functions from schema_tables
* database: modify_keyspace_on_all_shards: execute func first on shard 0 
   - implemented the ordering changes
* storage_service: keyspace_changed: execute only on shard 0
* migration_manager: propagate listener notification exceptions
*  database: update_keyspace_on_all_shards: abort on error
  - Ref  #3389

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1705/

bhalevy commented:
rebased

bhalevy commented:
@tgrabiec please review / merge.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1966/

tgrabiec commented on replica/database.cc:
```diff
@@ -896,12 +897,16 @@ future<> database::update_keyspace(sharded<service::storage_proxy>& proxy, const
+        });
+    } catch (...) {
+        on_fatal_internal_error(dblog, fmt::format("Failed updating keyspace {}: {}", name, std::current_exception()));
+    }
```
> 
> We don't do that for create/drop keyspace since
> this affects only that keyspace that is either not used yet
> (on create) or not used anymore (in the drop case).
> FWIW, the error is now propagated upstream.

But this still leaves the node in a bad state, which for example will make the recreation with the same name fail, if we ignore the error. But do we really ignore it? The exception will propagate to the group0 apply fiber and halt the group0 state machine. In that case, catching exceptions here is also unnecessary.

bhalevy replied:
okay.  I wasn't aware of the group0 handling.
That said, if we decided to backport this series, older branches may need this catch here.

bhalevy replied:
dropped this try/catch in v4

bhalevy commented on replica/database.cc:
```diff
@@ -1422,7 +1422,18 @@ future<> database::create_in_memory_keyspace(const lw_shared_ptr<keyspace_metada
+    // Run func first on shard 0
+    // to allow "seeding" of the effective_replication_map
+    // with a new e_r_m instance.
+    co_await sharded_db.invoke_on(0, func);
```
> > It is more efficient to do that first on shard 0
> > and then on all other shards, otherwise multiple
> > shards might need to calculate to new e_r_m (and reach
> > the same result).  When the new e_r_m is "seeded" on
> > shard 0, other shards will find it there and clone
> > a local copy of it - which is ore efficient.
> 
> Why can't we rely on listeners to do that?

To do what?  invoke_on shard 0?
We can, but that will just replicate the logic in a lower level.
since `modify_keyspace_on_all_shards` is defined as private I think it makes sense to centralize the logic in a middle layer function.

> keyspace::create_replication_strategy() already has a preemption point between _replication_strategy is assigned and update_effective_replication_map(). So users may observe inconsistent state. They could just as well wait a bit longer until all shards have updated metadata. Or is that an unintended bug?

It is unintended.

bhalevy replied:
kept `modify_keyspace_on_all_shards` in v4

tgrabiec replied:
> To do what?  invoke_on shard 0?

I meant to replicate erm. The listener could replicate erm, in which case we can ditch _erm_factory and not special-case shard 0 here. But if we need to change _replication_strategy atomically, that won't do.

bhalevy commented on replica/database.cc:
```diff
@@ -1422,7 +1422,18 @@ future<> database::create_in_memory_keyspace(const lw_shared_ptr<keyspace_metada
+    // to allow "seeding" of the effective_replication_map
+    // with a new e_r_m instance.
+    co_await sharded_db.invoke_on(0, func);
+    if (smp::count > 1) {
```
Hmm, yeah, maybe it's not worth optimizing for since it is most likely true and in the cases that it isn't, the penalty is minuscule with code optimizations.

bhalevy replied:
dropped in v4

bhalevy commented on replica/database.cc:
```diff
@@ -1417,9 +1417,19 @@ future<> database::create_in_memory_keyspace(const lw_shared_ptr<keyspace_metada
-database::create_keyspace(const lw_shared_ptr<keyspace_metadata>& ksm, locator::effective_replication_map_factory& erm_factory) {
-    return create_keyspace(ksm, erm_factory, system_keyspace::no);
+database::create_keyspace(sharded<service::storage_proxy>& proxy, const sstring& name) {
+    auto val = co_await db::schema_tables::read_schema_partition_for_keyspace(proxy, db::schema_tables::KEYSPACES, name);
```
Well, this flow is preexisting for `database::update_keyspace`.
But I think separating the logic as you said does make sense.
I'll take a stab at it and call `read_schema_partition_for_keyspace` (and `extract_scylla_specific_keyspace_info`) once in the schema_tables layer and pass the result(s) to all shards.

bhalevy replied:
fixed in v4.
plus aligned update_keyspace on the same lines.

bhalevy commented on transport/event_notifier.cc:
```diff
@@ -43,11 +44,15 @@ void cql_server::event_notifier::on_create_keyspace(const sstring& ks_name)
                 ks_name
             }));
+          } catch (...) {
+            elogger.warn("Create keyspace notification failed {}: {}. Ignored", ks_name, std::current_exception());
```
I'm not even sure how to drop it.

bhalevy replied:
I see that we have `conn->shutdown()`.
But I suggest we do this as a followup since it may risk this PR that is a step forward in the right direction.

bhalevy replied:
done in v4

bhalevy commented:
In v4 ([8fc0595](https://github.com/scylladb/scylladb/commit/8fc05959eb5c633cc9afc141f507bc07884558d5)):
- schema_tables: merge_keyspaces: extract_scylla_specific_keyspace_info for update_keyspace
- database: modify_keyspace_on_all_shards: call notifiers only after applying func on all shards
  - refactored
- storage_service: keyspace_changed: execute only on shard 0
  - dropped optimization for smp::count==1
- migration_manager: propagate listener notification exceptions
  - Added migration_notifier::on_schema_change helper
  - moved event_notifier changes to next patch
- transport: event_notifier: shutdown connection on error
  - use event_notifier::on_schema_change helper (and a couple others on top of it)
  - shutdown connection on error as per https://github.com/scylladb/scylladb/pull/14158#discussion_r1240234643

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2068/

bhalevy commented:
In v5:
- fixed cross-shard access to `lw_shared_ptr<keyspace_metadata>` from `create_keyspace_on_all_shards`

tgrabiec commented on transport/event_notifier.cc:
```diff
@@ -38,53 +38,67 @@ void cql_server::event_notifier::unregister_connection(cql_server::connection* c
+        } catch (...) {
+            ex = std::current_exception();
+            elogger.warn("{}. Closing connection", describe_error(ex));
+            conns_to_shutdown.emplace_back(conn);
```
Looking at `cql_server::connection::write_response`, there is nothing which can fail there. The writing happens in the background. What can fail is construction of the message (conn->make_schema_change_event()). But it is more likely that boost::copy_range() above fails, which is not handled. 

Maybe this effort to handle failure here is futile after all. Why not let the exception from the listener propagate up like from other listeners, and fail the schema change?

bhalevy replied:
Makes sense.  Initially I thought we need to squelch the errors to preserve the existing behavior but you convinced me otherwise.
I'll drop this change then.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2082/

bhalevy commented:
In v6:
- dropped "transport: event_notifier: shutdown connection on error"

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2095/

bhalevy commented:
@tgrabiec please re-review

mykaul commented:
Do we need to backport this to 5.2 (and thus 2023.1) ?

bhalevy commented:
> Do we need to backport this to 5.2 (and thus 2023.1) ?

I don't think so.
It's not a regression and I doubt we ever emcountered the issue in the wild.

bhalevy commented:
Backport to 5.2 in https://github.com/scylladb/scylladb/pull/15872
